internal/debug/lifecycle.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2017-2021 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debug
+
+import "go.uber.org/atomic"
+
+type (
+	// lifeCycleImpl implements the LifeCycle interface
+	lifeCycleImpl struct {
+		pollerCount atomic.Int32
+	}
+
+	// runImpl implements the Run interface
+	runImpl struct {
+		lifeCycle *lifeCycleImpl
+		stopped   atomic.Bool
+	}
+
+	// Run is a helper for simpler tracking of the go routine run
+	// Deprecated: in development and very likely to change
+	Run interface {
+		// Stop is the method to report stats once a poller thread is done
+		Stop()
+	}
+
+	// LifeCycle contains a set of methods to collect information on a running worker
+	// Deprecated: in development and very likely to change
+	LifeCycle interface {
+		// PollerStart collects information on poller start up.
+		// consumers should provide a concurrency-safe implementation.
+		PollerStart(workerID string) Run
+		// ReadPollerCount return the number or running pollers
+		ReadPollerCount() int32
+	}
+
+	// EventMonitoring provides a set of methods that can be used to collect
+	// stats on the Worker for debugging purposes.
+	// Deprecated: in development and very likely to change
+	EventMonitoring struct {
+		LifeCycle LifeCycle
+	}
+)
+
+func (lc *lifeCycleImpl) PollerStart(workerID string) Run {
```

Comment:
```suggestion
func (lc *lifeCycleImpl) PollerStart() Run {
```
We are not using the argument here? Maybe we should have a map or something? But currently it's not used.

Reply:
The initial plan was to define the interface only and the implementation would need to be provided by users. Internally, we’d have only a noop implementation for backward compatibility. Having only a Start() and Stop() with no args would limit the users on their implementation. That’s why I added at least workerID as an argument. 
Eventually I added a default implementation and though it is not used in the default workerID just seemed like a good starting point for consumers implementation